lib/engines/gcp/gcp.ex

```diff
@@ -0,0 +1,279 @@
+defmodule Ptolemy.Engines.GCP do
+  @moduledoc """
+  `Ptolemy.Engines.GCP` provides interaction with Vault's Google Cloud Secrets Engine.
+
+  {{:app_name, :another_secret}, {GCP.Provider, {server, engine_name, gcp_secret_type, secret_name}}}
+
+  """
+  alias Ptolemy.Server
+  alias Ptolemy.Engines.GCP.Engine
+
+  @typedoc """
+  Types of Google Cloud Secrets allowed by Vault
+  """
+  @type gcp_secret_type() :: :access_token | :service_account_key
+
+  @typedoc """
+    A GCP roleset map
+
+    ## Fields
+    * `:secret_type` - type of secret generated for this role set. i.e. "access_token", "service_account_key"
+    * `:project` - name of the GCP project to which this roleset's service account will belong
+    * `:bindings` - bindings configuration string (read more here: https://www.vaultproject.io/docs/secrets/gcp/index.html#roleset-bindings)
+    * `:token_scopes` - *Applies only if secret type is `access_token`* list of OAuth scopes belonging to secrets generated under this role set
+  """
+  @type roleset :: %{
+          required(:secret_type) => String.t(),
+          required(:project) => String.t(),
+          required(:bindings) => String.t(),
+          optional(:token_scopes) => list(String.t())
+        }
+
+  @doc """
+  Creates a roleset account in the given engine and returns `:ok` if everything goes well.
+  Throws an error message otherwise.
+  """
+  @spec create!(pid(), atom(), String.t(), roleset) :: :ok
+  def create!(pid, engine_name, roleset_name, roleset_payload) do
+    create(pid, engine_name, roleset_name, roleset_payload)
+    |> case do
+      {:ok, _body} -> :ok
+      {:error, err} -> throw(err)
```

Comment:
Raise instead of throwing would be more appropriate here, since bang functions usuallly raises exceptions instead of throwing values. We only throw when we try-catch to pass a value that cannot be passed in any other ways.

Reply:
👍 fixed in commit "throw -> raise"